### PR TITLE
Reland "Enable GRPC_ERROR_IS_ABSEIL_STATUS"

### PR DESCRIPTION
**Pull request by veblush**
Roll-forward https://github.com/grpc/grpc/pull/29869
Revert https://github.com/grpc/grpc/pull/30031

This will be cherry-picked via cl/455649608

### CODE REVIEW

File: test/cpp/server/server_request_call_test.cc [hunk 1]
```diff
@@ -135,7 +135,15 @@ TEST(ServerRequestCallTest, ShortDeadlineDoesNotCauseOkayFalse) {
     ctx.set_deadline(std::chrono::system_clock::now() +
                      std::chrono::milliseconds(1));
     grpc::Status status = stub->Echo(&ctx, request, &response);
-    EXPECT_EQ(StatusCode::DEADLINE_EXCEEDED, status.error_code());
+    if (i == 0 && status.error_code() == grpc::OK) {
+      // This first call may take more time to trigger a deadline excceded
```
Comment (markdroth) [inline test/cpp/server/server_request_call_test.cc:139]: Why does this happen, and what does this have to do with the change to use `absl::Status` instead of `grpc_error`?

Comment (ctiller) [reply test/cpp/server/server_request_call_test.cc:139]: also interested here!

Comment (veblush) [reply test/cpp/server/server_request_call_test.cc:139]: As the comment say, it's because it takes longer time in Bazel-RBE environment (which doesn't have a strong machine to run the test) for the first call. This is because absl::Status uses a heap unlike grpc_error (which uses bespoke arena) and it has to take longer time to initialize it.

Comment (markdroth) [reply test/cpp/server/server_request_call_test.cc:139]: I don't understand that explanation.  Regardless of how long it takes to allocate the `absl::Status`, that should not affect the outcome of this test.  What exact sequence of events is leading to this outcome?  Where is the `absl::Status` being allocated, and why does that prevent the deadline timer from firing on time?

In principle, I think it should not be possible for this to get an OK status, because the server's response will never come before the deadline is exceeded.  If we are somehow returning OK in a case where we should be returning DEADLINE_EXCEEDED due to the `absl::Status` not being initialized when it should be, then this is a fairly serious bug, and we need to fix that bug instead of changing the test to accept the buggy output.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: test/cpp/server/server_request_call_test.cc (hunk 1 above)
Comment (veblush) [reply test/cpp/server/server_request_call_test.cc:139]: This is how it's supposed to happen.

1. Client sends a call with 1ms deadline. (Set a timer for a deadline)
2. Sever receives a call and waits for 50ms and returns OK.
3. Deadline timer fires and passing DEADLINE_EXCEEDED. 
4. Client receives DEADLINE_EXCEEDED.

If the 3rd step takes longer (> 50ms), client will receive OK. That's why it happens to have OK. Again this test is pretty sensitive to timing and RBE environment doesn't assign a strong machine to it. `absl::Status` takes longer time when creating non-OK status (one of reasons is to use a heap) and heap needs time to expand in this case to store additional metadata in the status.

Comment (ctiller) [reply test/cpp/server/server_request_call_test.cc:139]: If the timeouts are wrong for our environment I'd strongly suggest adjusting the timeouts rather than weakening the test.

(i.e. leave the test logic as is, maybe increase the time for step 2 to several seconds?)

Comment (veblush) [reply test/cpp/server/server_request_call_test.cc:139]: Yeah it would work too. Having a longer delay on the service side worked so it could be a solution. Just want to make sure that having a shorter delay (50ms) here is required here or not. @markdroth

Comment (markdroth) [reply test/cpp/server/server_request_call_test.cc:139]: I agree that increasing the delay is a better solution.  The purpose of this test is to show that a short deadline does not cause ok=false.  We can't prove that if the deadline timer does not fire before the RPC completes.
